packages/libs/deer/src/error/type.rs

```diff
@@ -0,0 +1,179 @@
+use alloc::format;
+use core::fmt::{Display, Formatter};
+
+use super::{Error, ErrorProperties, ErrorProperty, Id, Namespace, NAMESPACE};
+use crate::{
+    error::{macros::impl_error, Location, Schema},
+    id,
+};
+
+#[derive(serde::Serialize)]
+pub struct ExpectedType(Schema);
+
+impl ExpectedType {
+    #[must_use]
+    pub const fn new(schema: Schema) -> Self {
+        Self(schema)
+    }
+
+    pub(crate) const fn schema(&self) -> &Schema {
+        &self.0
+    }
+}
+
+impl ErrorProperty for ExpectedType {
+    type Value<'a> = Option<&'a Self>;
+
+    fn key() -> &'static str {
+        "expected"
+    }
+
+    fn value<'a>(mut stack: impl Iterator<Item = &'a Self>) -> Self::Value<'a> {
+        stack.next()
+    }
+}
+
+#[derive(serde::Serialize)]
+pub struct ReceivedType(Schema);
+
+impl ReceivedType {
+    #[must_use]
+    pub const fn new(schema: Schema) -> Self {
+        Self(schema)
+    }
+
+    pub(crate) const fn schema(&self) -> &Schema {
+        &self.0
+    }
+}
+
+impl ErrorProperty for ReceivedType {
+    type Value<'a> = Option<&'a Self>;
+
+    fn key() -> &'static str {
+        "received"
+    }
+
+    fn value<'a>(mut stack: impl Iterator<Item = &'a Self>) -> Self::Value<'a> {
+        stack.next()
+    }
```

Comment:
Isn't this just the same as `Iterator::next`?

Reply:
Yes, but that isn't always the case. Some implementations might want to take into account other values. For example, `Location` renders a list of all items, but one could also imagine the possibility of future types of filtering items, etc.